File: src/main/java/uk/gov/justice/digital/hmpps/riskprofiler/events/EventListener.java
package uk.gov.justice.digital.hmpps.riskprofiler.events;

import com.fasterxml.jackson.databind.ObjectMapper;
import lombok.extern.slf4j.Slf4j;
import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
import org.springframework.jms.annotation.JmsListener;
import org.springframework.stereotype.Service;
import uk.gov.justice.digital.hmpps.riskprofiler.model.OffenderEvent;
import uk.gov.justice.digital.hmpps.riskprofiler.services.NomisService;
import uk.gov.justice.digital.hmpps.riskprofiler.services.PollPrisonersService;

import java.io.IOException;
import java.util.Map;

@Service
@ConditionalOnProperty(name = "sqs.provider")
@Slf4j
public class EventListener {

    private final NomisService nomisService;
    private final PollPrisonersService pollPrisonersService;
    private final ObjectMapper objectMapper;

    public EventListener(final NomisService nomisService, final PollPrisonersService pollPrisonersService, final ObjectMapper objectMapper) {
        this.nomisService = nomisService;
        this.pollPrisonersService = pollPrisonersService;
        this.objectMapper = objectMapper;
    }

    @JmsListener(destination = "${sqs.events.queue.name}")
    public void eventListener(final String requestJson) {
        final var event = getOffenderEvent(requestJson);
        if (event != null) {
            switch (event.getEventType()) {
                case "ALERT-INSERTED":
                case "ALERT-UPDATED":
                case "ALERT-DELETED":
                    final var isEscape = NomisService.ESCAPE_LIST_ALERT_TYPES.contains(event.getAlertCode());
                    final var isSoc = NomisService.SOC_ALERT_TYPES.contains(event.getAlertCode());
                    if ((isEscape || isSoc) && event.getBookingId() != null) {
                        // booking id can be null for ALERT-DELETED. In this case ignore, as it is unlikely it will trigger need to change to cat B
                        final var nomsId = nomisService.getOffender(event.getBookingId());
                        if (isEscape) {
                            nomisService.evictEscapeListAlertsCache(nomsId);
                        }
                        if (isSoc) {
                            nomisService.evictSocListAlertsCache(nomsId);
                        }
                        pollPrisonersService.pollPrisoner(nomsId);
                    }
                    break;
                case "INCIDENT-INSERTED":
                case "INCIDENT-CHANGED-CASES":
                case "INCIDENT-CHANGED-PARTIES":
                case "INCIDENT-CHANGED-RESPONSES":
                case "INCIDENT-CHANGED-REQUIREMENTS":
                    final var nomsIds = nomisService.getPartiesOfIncident(event.getIncidentCaseId());
                    // TODO: Do not poll immediately (leave for batch) as one incident generates a lot of events
                    // Also should apply filter choosing only the relevant types of Assault incidents
                    nomsIds.forEach(nomisService::evictIncidentsCache);
                    break;
            }
        }
    }

    private OffenderEvent getOffenderEvent(final String requestJson) {
        OffenderEvent event = null;
        try {
            final Map<String, String> message = objectMapper.readValue(requestJson, Map.class);
            if (message == null || message.get("Message") == null) {
                log.warn(requestJson);
            } else {
                log.debug(message.get("Message")); // do not log an excessive amount of data
                event = objectMapper.readValue(message.get("Message"), OffenderEvent.class);
            }
        } catch (final IOException e) {
            log.error("Failed to Parse Message {}", requestJson);
        }
        return event;
    }
}

File: src/main/kotlin/uk/gov/justice/digital/hmpps/riskprofiler/events/EventListener.kt
package uk.gov.justice.digital.hmpps.riskprofiler.events

import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.module.kotlin.readValue
import org.slf4j.LoggerFactory
import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty
import org.springframework.jms.annotation.JmsListener
import org.springframework.stereotype.Service
import uk.gov.justice.digital.hmpps.riskprofiler.model.OffenderEvent
import uk.gov.justice.digital.hmpps.riskprofiler.services.NomisService
import uk.gov.justice.digital.hmpps.riskprofiler.services.PollPrisonersService
import java.io.IOException
import java.util.function.Consumer

@Service
@ConditionalOnProperty(name = ["sqs.provider"])
class EventListener(
  private val nomisService: NomisService,
  private val pollPrisonersService: PollPrisonersService,
  private val objectMapper: ObjectMapper,
) {
  @JmsListener(destination = "\${sqs.events.queue.name}")
  fun eventListener(requestJson: String) {
    val event = getOffenderEvent(requestJson)
    if (event != null) {
      when (event.eventType) {
        "ALERT-INSERTED", "ALERT-UPDATED", "ALERT-DELETED" -> {
          val isEscape = NomisService.ESCAPE_LIST_ALERT_TYPES.contains(event.alertCode)
          val isSoc = NomisService.SOC_ALERT_TYPES.contains(event.alertCode)
          if ((isEscape || isSoc) && event.bookingId != null) {
            // booking id can be null for ALERT-DELETED. In this case ignore, as it is unlikely it will trigger need to change to cat B
            val nomsId = nomisService.getOffender(event.bookingId)!!
            if (isEscape) {
              nomisService.evictEscapeListAlertsCache(nomsId)
            }
            if (isSoc) {
              nomisService.evictSocListAlertsCache(nomsId)
            }
            pollPrisonersService.pollPrisoner(nomsId)
          }
        }
        "INCIDENT-INSERTED", "INCIDENT-CHANGED-CASES", "INCIDENT-CHANGED-PARTIES", "INCIDENT-CHANGED-RESPONSES", "INCIDENT-CHANGED-REQUIREMENTS" -> {
          val nomsIds = nomisService.getPartiesOfIncident(event.incidentCaseId)
          // TODO: Do not poll immediately (leave for batch) as one incident generates a lot of events
          // Also should apply filter choosing only the relevant types of Assault incidents
          nomsIds.forEach(Consumer { nomsId: String? -> nomisService.evictIncidentsCache(nomsId) })
        }
      }
    }
  }

  private fun getOffenderEvent(requestJson: String): OffenderEvent? {
    var event: OffenderEvent? = null
    try {
      val message: Map<String, Any?> = objectMapper.readValue(requestJson)
      if (message["Message"] == null) {
        log.warn(requestJson)
      } else {
        val body = message["Message"] as String
        log.debug(body) // do not log an excessive amount of data
        event = objectMapper.readValue(body)
      }
    } catch (e: IOException) {
      log.error("Failed to Parse Message {} {}", requestJson, e)
    }
    return event
  }

  companion object {
    private val log = LoggerFactory.getLogger(EventListener::class.java)
  }
}
